build/source/engine/allocspace.f90

```diff
@@ -19,405 +19,348 @@
- nVar = size(attr_meta)
- do iHRU=1,nHRU
-  allocate(attr_hru(iHRU)%var(nVar),stat=err)
-  if(err/=0)then; err=20; message=trim(message)//"problemAllocateData2ndLevel"; return; endif
-  ! fill data with missing values
-  attr_hru(iHRU)%var(:) = missingDouble
- end do
- end subroutine alloc_attr
+ ! input-output
+ type(var_info),intent(in)        :: metaStruct(:)  ! metadata structure
+ class(*),intent(inout)           :: dataStruct     ! data structure
+ ! optional input
+ integer(i4b),intent(in),optional :: nSnow          ! number of snow layers
+ integer(i4b),intent(in),optional :: nSoil          ! number of soil layers
+ ! output
+ integer(i4b),intent(out)         :: err            ! error code
+ character(*),intent(out)         :: message        ! error message
+ ! local
+ integer(i4b)                     :: iVar           ! loop through variables in the metadata structure
+ logical(lgt)                     :: check          ! .true. if the variables are allocated
+ integer(i4b)                     :: nVars          ! number of variables in the metadata structure
+ integer(i4b)                     :: nLayers        ! total number of layers
+ character(len=256)               :: cmessage       ! error message of the downwind routine
+ ! initialize error control
+ err=0; message='allocLocal/'
 
+ ! get the number of variables in the metadata structure
+ nVars = size(metaStruct)
 
- ! *************************************************************************************************
- ! public subroutine alloc_type: initialize data structures for local classification of veg, soil, etc.
- ! *************************************************************************************************
- subroutine alloc_type(nHRU,err,message)
- ! used to initialize structure components for model variables
- USE data_struc,only:type_hru,type_meta             ! data structures
- implicit none
- ! dummy variables
- integer(i4b),intent(in)              :: nHRU        ! number of HRUs
- integer(i4b),intent(out)             :: err         ! error code
- character(*),intent(out)             :: message     ! error message
- ! local variables
- integer(i4b)                         :: iHRU        ! loop through HRUs
- integer(i4b)                         :: nVar        ! number of variables
- ! initialize errors
- err=0; message="f-alloc_type/"
- ! check that the metadata structure is allocated
- if(.not.associated(type_meta))then
-  err=10; message=trim(message)//"metadataNotInitialized"; return
+ ! check if nSnow and nSoil are present
+ if(present(nSnow) .or. present(nSoil))then
+  ! check both are present
+  if(.not.present(nSoil))then; err=20; message=trim(message)//'expect nSoil to be present when nSnow is present'; return; endif
+  if(.not.present(nSnow))then; err=20; message=trim(message)//'expect nSnow to be present when nSoil is present'; return; endif
+  nLayers = nSnow+nSoil
+
+ ! check that nSnow and nSoil are not needed
```

Comment:
It is possible that `nSnow` and `nSoil` are actually needed here, so we return an error if the optional arguments are missing when needed.

Will revise the comment.